plugins/dell-dock2/fu-dell-dock2-rtshub.c

```diff
@@ -0,0 +1,586 @@
+	    .ext = RTSHUB_EXT_RESET_TO_FLASH,
+	    .dwregaddr = 0,
+	    .bufferlen = 0,
+	};
+	guint8 buf[DELL_DOCK2_RTSHUB_BUFFER_SIZE] = {0};
+
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    0x0, /* dst */
+			    (const guint8 *)&cmd_buffer,
+			    sizeof(cmd_buffer),
+			    0x0, /* src */
+			    sizeof(cmd_buffer),
+			    error))
+		return FALSE;
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      0x0,
+				      buf,
+				      sizeof(buf),
+				      DELL_DOCK2_RTSHUB_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "failed to soft reset: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_rtshub_write_flash(FuDellDock2RtsHub *self,
+				 guint32 addr,
+				 const guint8 *data,
+				 guint16 data_sz,
+				 GError **error)
+{
+	FuRtshubHIDCmdBuffer cmd_buffer = {
+	    .cmd = RTSHUB_CMD_WRITE_DATA,
+	    .ext = RTSHUB_EXT_WRITEFLASH,
+	    .dwregaddr = GUINT32_TO_LE(addr),
+	    .bufferlen = GUINT16_TO_LE(data_sz),
+	};
+	guint8 buf[DELL_DOCK2_RTSHUB_BUFFER_SIZE] = {0};
+
+	g_return_val_if_fail(data_sz <= 128, FALSE);
+	g_return_val_if_fail(data != NULL, FALSE);
+	g_return_val_if_fail(data_sz != 0, FALSE);
+
+	/* vendor command */
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    0x0, /* dst */
+			    (const guint8 *)&cmd_buffer,
+			    sizeof(cmd_buffer),
+			    0x0, /* src */
+			    sizeof(cmd_buffer),
+			    error))
+		return FALSE;
+
+	/* data payload */
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    DELL_DOCK2_RTSHUB_WRITE_FLASH_OFFSET_DATA, /* dst */
+			    data,
+			    data_sz,
+			    0x0, /* src */
+			    data_sz,
+			    error))
+		return FALSE;
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      0x0,
+				      buf,
+				      sizeof(buf),
+				      DELL_DOCK2_RTSHUB_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "failed to write flash @%08x: ", (guint)addr);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gchar *
+fu_dell_dock2_rtshub_substr_before_colon(const gchar *str)
+{
+	gchar *colon = g_strrstr(str, ":");
+	return (colon != NULL) ? g_strndup(str, colon - str) : g_strdup(str);
+}
+
+static gboolean
+fu_dell_dock2_rtshub_is_in_family(FuDevice *device, FuDevice *sibling)
+{
+	const gchar *dev_phyid = fu_device_get_physical_id(device);
+	const gchar *sib_phyid = fu_device_get_physical_id(sibling);
+	const gchar *dev_phyid_prefix = fu_dell_dock2_rtshub_substr_before_colon(dev_phyid);
+
+	g_debug("device (%s) physical id: %s", fu_device_get_name(sibling), sib_phyid);
+
+	/* compare physical id */
+	return g_str_has_prefix(sib_phyid, dev_phyid_prefix);
+}
+
+static gboolean
+fu_dell_dock2_rtshub_care_family_back(FuDevice *device, GError **error)
+{
+	FuDellDock2RtsHub *self = FU_DELL_DOCK2_RTSHUB(device);
+	FuDevice *parent = fu_device_get_parent(device);
+
+	/* wait myself */
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
```

Comment:
@hughsie I've added a flag for 413C:B0A1 to come back, for some reason the daemon removes it.  Do you konw how to figure out the reason it is being removed?

[rts5480_gen1_device_removed.log](https://github.com/user-attachments/files/16539760/rts5480_gen1_device_removed.log)

```
08:03:38.030 FuDeviceList         doing delayed removal
08:03:38.030 FuDeviceList         ::removed f40fdba32387902e0b799c11d27e2382603bd22d [RTS5480 Gen 1 USB Hub]
08:03:38.030 FuPlugin             backend_device_removed(dell_dock)
08:03:38.030 FuPlugin             backend_device_removed(linux_display)
08:03:38.031 FuMain               REMOVED:
FuDellDock2RtsHub:
  DeviceId:             f40fdba32387902e0b799c11d27e2382603bd22d
  CompositeId:          acdfe8b03e2cfefbc9f0d4680595549d0a0e9c4a
  Name:                 RTS5480 Gen 1 USB Hub
  Status:               device-busy
  Percentage:           100
  Guid:                 9d84dbd7-8750-5c6b-8e55-57cfd9f88fda <- USB\VID_413C&PID_B0A1
  Guid:                 cbe221f0-246a-53fa-923b-25d05aa2d240 <- USB\VID_413C&PID_B0A1&DOCKTYPE_07
  Plugin:               dell_dock2
  Protocol:             com.dell.dock2
  Flags:                registered|wait-for-replug|updatable-hidden|signed-payload
  Problems:             update-in-progress
```

Reply:
Is fu_device_set_remove_delay() being set in _init()?

Reply:
> Is fu_device_set_remove_delay() being set in _init()?

Yes, it's set in device init.